diff -urN old/src/BulletDynamics/Character/btCharacterControllerInterface.h new/src/BulletDynamics/Character/btCharacterControllerInterface.h
--- old/src/BulletDynamics/Character/btCharacterControllerInterface.h	2011-04-11 12:39:50.000000000 -0500
+++ new/src/BulletDynamics/Character/btCharacterControllerInterface.h	2011-04-11 12:40:29.000000000 -0500
@@ -31,7 +31,7 @@
 	
 	virtual void	setWalkDirection(const btVector3& walkDirection) = 0;
 	virtual void	setVelocityForTimeInterval(const btVector3& velocity, btScalar timeInterval) = 0;
-	virtual void	reset () = 0;
+	virtual void	reset ( btCollisionWorld* collisionWorld ) = 0;
 	virtual void	warp (const btVector3& origin) = 0;
 
 	virtual void	preStep ( btCollisionWorld* collisionWorld) = 0;
diff -urN old/src/BulletDynamics/Character/btKinematicCharacterController.cpp new/src/BulletDynamics/Character/btKinematicCharacterController.cpp
--- old/src/BulletDynamics/Character/btKinematicCharacterController.cpp	2011-04-11 12:39:50.000000000 -0500
+++ new/src/BulletDynamics/Character/btKinematicCharacterController.cpp	2011-04-11 12:40:57.000000000 -0500
@@ -467,8 +467,21 @@
 
 
 
-void btKinematicCharacterController::reset ()
+void btKinematicCharacterController::reset ( btCollisionWorld* collisionWorld )
 {
+	m_verticalVelocity = 0.0;
+        m_verticalOffset = 0.0;
+        m_wasOnGround = false;
+        m_wasJumping = false;
+        m_walkDirection.setValue(0,0,0);
+        m_velocityTimeInterval = 0.0;
+
+        //clear pair cache
+        btHashedOverlappingPairCache *cache = m_ghostObject->getOverlappingPairCache();
+        while (cache->getOverlappingPairArray().size() > 0)
+        {
+                cache->removeOverlappingPair(cache->getOverlappingPairArray()[0].m_pProxy0, cache->getOverlappingPairArray()[0].m_pProxy1, collisionWorld->getDispatcher());
+        }
 }
 
 void btKinematicCharacterController::warp (const btVector3& origin)
diff -urN old/src/BulletDynamics/Character/btKinematicCharacterController.h new/src/BulletDynamics/Character/btKinematicCharacterController.h
--- old/src/BulletDynamics/Character/btKinematicCharacterController.h	2011-04-11 12:39:50.000000000 -0500
+++ new/src/BulletDynamics/Character/btKinematicCharacterController.h	2011-04-11 12:40:29.000000000 -0500
@@ -129,7 +129,7 @@
 	virtual void setVelocityForTimeInterval(const btVector3& velocity,
 				btScalar timeInterval);
 
-	void reset ();
+	void reset ( btCollisionWorld* collisionWorld );
 	void warp (const btVector3& origin);
 
 	void preStep (  btCollisionWorld* collisionWorld);
