--- old/src/BulletDynamics/Character/btKinematicCharacterController.cpp.old  2010-07-21 16:33:40.000000000 -0500
+++ new/src/BulletDynamics/Character/btKinematicCharacterController.cpp	2011-02-04 10:59:48.000000000 -0600
@@ -393,10 +393,12 @@
 	m_targetPosition -= (step_drop + gravity_drop);*/
 
 	btScalar downVelocity = (m_verticalVelocity<0.f?-m_verticalVelocity:0.f) * dt;
-	if(downVelocity > 0.0 && downVelocity < m_stepHeight
+	//if(downVelocity > 0.0 && downVelocity < m_stepHeight
+	if(downVelocity > m_fallSpeed
 		&& (m_wasOnGround || !m_wasJumping))
 	{
-		downVelocity = m_stepHeight;
+		//downVelocity = m_stepHeight;
+		downVelocity = m_fallSpeed;
 	}
 
 	btVector3 step_drop = getUpAxisDirections()[m_upAxis] * (m_currentStepOffset + downVelocity);
