diff -urN CS.old/configure CS.new/configure
--- CS.old/configure	2008-04-22 07:14:51.000000000 -0500
+++ CS.new/configure	2009-04-13 14:26:31.000000000 -0500
@@ -67509,6 +67509,8 @@
 	[Ii][3-9]86*|[Xx]86*) cs_host_cpu=x86 ;;
 	powerpc*) cs_host_cpu=powerpc ;;
 	sparc*) cs_host_cpu=sparc ;;
+	mips*) cs_host_cpu=mips ;;
+	alpha*) cs_host_cpu=alpha ;;
 	*) cs_host_cpu=$host_cpu ;;
     esac
     cs_host_cpu_normalized="`echo "$cs_host_cpu" | $as_tr_cpp`"
@@ -67530,6 +67532,17 @@
 fi
 
 
+if test $cs_host_cpu = sparc; then
+  cs_build_prop_val="-mcpu=v9"
+    cs_build_prop_val=`echo x$cs_build_prop_val | sed 's/^x//;s/   */ /g;s/^ //;s/ $//'`
+    if test -n "$cs_build_prop_val"; then
+  cs_jamconfig_text="${cs_jamconfig_text}COMPILER.C++FLAGS += \"$cs_build_prop_val\" ;
+"
+fi
+
+fi
+
+
 
 
 #------------------------------------------------------------------------------
diff -urN CS.old/configure.ac CS.new/configure.ac
--- CS.old/configure.ac	2008-04-22 07:20:30.000000000 -0500
+++ CS.new/configure.ac	2009-04-13 14:25:01.000000000 -0500
@@ -1811,6 +1811,8 @@
 	[[Ii][3-9]86*|[Xx]86*]) cs_host_cpu=x86 ;;
 	powerpc*) cs_host_cpu=powerpc ;;
 	sparc*) cs_host_cpu=sparc ;;
+	mips*) cs_host_cpu=mips ;;
+	alpha*) cs_host_cpu=alpha ;;
 	*) cs_host_cpu=$host_cpu ;;
     esac
     cs_host_cpu_normalized="AS_TR_CPP([$cs_host_cpu])"])
@@ -1824,6 +1826,9 @@
 AS_IF([test $cs_host_cpu = x86], 
   [CS_HEADER_PROPERTY([CS_HAVE_MMX])])
 
+AS_IF([test $cs_host_cpu = sparc],
+  [CS_EMIT_BUILD_PROPERTY([COMPILER.C++FLAGS], [-mcpu=v9], [append])])
+
 
 
 #------------------------------------------------------------------------------
