diff -urN CS.old/plugins/mesh/genmesh/object/submeshes.cpp CS/plugins/mesh/genmesh/object/submeshes.cpp
--- CS.old/plugins/mesh/genmesh/object/submeshes.cpp	2010-08-31 18:31:58.000000000 -0500
+++ CS/plugins/mesh/genmesh/object/submeshes.cpp	2010-08-31 18:32:46.000000000 -0500
@@ -165,11 +165,11 @@
   static int SubmeshSubmeshCompare (SubMesh* const& A, 
                                     SubMesh* const& B)
   {
-    const char* a = A->GetName();
-    const char* b = B->GetName();
-    if (a == 0) return (b == 0) ? 0 : 1;
-    if (b == 0) return -1;
-    return strcmp (a, b);
+    if (A < B)
+    	return -1;
+    if (A > B)
+    	return 1;
+    return 0;
   }
 
   iGeneralMeshSubMesh* SubMeshesContainer::AddSubMesh (
