--- CS/plugins/mesh/genmesh/object/submeshes.cpp	2008-05-27 17:35:08.000000000 -0500
+++ CS.new/plugins/mesh/genmesh/object/submeshes.cpp	2010-10-08 21:36:27.788247594 -0500
@@ -165,11 +165,11 @@
   static int SubmeshSubmeshCompare (SubMesh* const& A, 
                                     SubMesh* const& B)
   {
-    const char* a = A->GetName();
-    const char* b = B->GetName();
-    if (a == 0) return (b == 0) ? 0 : 1;
-    if (b == 0) return -1;
-    return strcmp (a, b);
+    if (A < B)
+    	return -1;
+    if (A > B)
+    	return 1;
+    return 0;
   }
 
   iGeneralMeshSubMesh* SubMeshesContainer::AddSubMesh (
@@ -217,6 +217,7 @@
       csArrayCmp<SubMesh*, SubMesh*> (subMesh, &SubmeshSubmeshCompare));
     if (idx == csArrayItemNotFound) return;
     subMeshes.DeleteIndex (idx);
+    changeNum++;
 
     if (subMeshes.GetSize() == 0)
       subMeshes.Push (defaultSubmesh);
