diff -urN orig/libs/csplugincommon/opengl/glfontcache.cpp new/libs/csplugincommon/opengl/glfontcache.cpp
--- orig/libs/csplugincommon/opengl/glfontcache.cpp	2009-09-09 13:39:03.078217495 -0500
+++ new/libs/csplugincommon/opengl/glfontcache.cpp	2009-09-09 12:10:29.905291000 -0500
@@ -651,7 +651,7 @@
   KnownFont* knownFont = GetCachedFont (font);
   if (knownFont == 0) knownFont = CacheFont (font);
 
-  if (pen_y <= ClipY1) return;
+  if (pen_y <= ClipY1 && G2D->fbHeight == G2D->vpHeight) return;
   pen_y = G2D->vpHeight - pen_y/* - maxheight*/;
 
   size_t textLen = isWide ? wcslen ((wchar_t*)text) : strlen ((char*)text);
