diff -urN CS.orig/include/ivideo/texture.h CS/include/ivideo/texture.h
--- CS.orig/include/ivideo/texture.h	2008-01-29 10:32:32.000000000 -0600
+++ CS/include/ivideo/texture.h	2010-10-17 01:16:16.720249844 -0500
@@ -31,6 +31,7 @@
 #include "csutil/scf_interface.h"
 #include "cstypes.h"
 #include "ivideo/graph3d.h"
+#include "igraphic/image.h"
 
 struct iGraphics2D;
 struct iGraphics3D;
@@ -287,6 +288,9 @@
     natureDirect = 1
   };
   virtual BlitBufferNature GetBufferNature (uint8* buf) = 0;
+
+  /// Dump the contents onto an image.
+  virtual csPtr<iImage> Dump () = 0;
 };
 
 /** @} */
diff -urN CS.orig/plugins/video/render3d/null/null_txt.h CS/plugins/video/render3d/null/null_txt.h
--- CS.orig/plugins/video/render3d/null/null_txt.h	2008-01-29 10:32:32.000000000 -0600
+++ CS/plugins/video/render3d/null/null_txt.h	2010-10-17 01:17:21.452396013 -0500
@@ -88,6 +88,9 @@
   }
   void ApplyBlitBuffer (uint8* buf) { cs_free (buf); }
   BlitBufferNature GetBufferNature (uint8* buf) { return natureDirect; }
+
+  /// Dump the contents onto an image (not currently implemented)
+  virtual csPtr<iImage> Dump () {return 0;}
 };
 
 #include "csutil/deprecated_warn_on.h"
diff -urN CS.orig/plugins/video/render3d/opengl/gl_txtmgr_basictex.h CS/plugins/video/render3d/opengl/gl_txtmgr_basictex.h
--- CS.orig/plugins/video/render3d/opengl/gl_txtmgr_basictex.h	2008-02-19 05:36:59.000000000 -0600
+++ CS/plugins/video/render3d/opengl/gl_txtmgr_basictex.h	2010-10-17 01:14:53.628246992 -0500
@@ -345,7 +345,7 @@
   BlitBufferNature GetBufferNature (uint8* buf);
 
   /// Dump the contents onto an image.
-  csPtr<iImage> Dump ();
+  virtual csPtr<iImage> Dump ();
 };
 
 #include "csutil/deprecated_warn_on.h"
diff -urN CS.orig/plugins/video/render3d/software/soft_txt.h CS/plugins/video/render3d/software/soft_txt.h
--- CS.orig/plugins/video/render3d/software/soft_txt.h	2008-01-29 10:32:32.000000000 -0600
+++ CS/plugins/video/render3d/software/soft_txt.h	2010-10-17 01:16:53.572275517 -0500
@@ -201,6 +201,9 @@
     size_t& pitch, TextureBlitDataFormat format, uint bufFlags);
   void ApplyBlitBuffer (uint8* buf);
   BlitBufferNature GetBufferNature (uint8* buf) { return natureDirect; }
+
+  /// Dump the contents onto an image (not currently implemented)
+  virtual csPtr<iImage> Dump () {return 0;}
 };
 
 #include "csutil/deprecated_warn_on.h"
