diff -urN CS/include/csgeom/poly3d.h CS.new/include/csgeom/poly3d.h
--- CS/include/csgeom/poly3d.h	2007-02-02 13:55:52.000000000 -0600
+++ CS.new/include/csgeom/poly3d.h	2009-11-06 15:31:51.000000000 -0600
@@ -351,7 +351,7 @@
 struct csCompressVertex
 {
   size_t orig_idx;
-  int x, y, z;
+  float x, y, z;
   size_t new_idx;
   bool used;
 };
diff -urN CS/libs/csgeom/poly3d.cpp CS.new/libs/csgeom/poly3d.cpp
--- CS/libs/csgeom/poly3d.cpp	2007-02-01 20:02:34.000000000 -0600
+++ CS.new/libs/csgeom/poly3d.cpp	2009-11-06 15:32:26.000000000 -0600
@@ -884,9 +884,9 @@
   for (i = 0; i < num_vertices; i++)
   {
     vt[i].orig_idx = i;
-    vt[i].x = (int)ceil (vertices[i].x * 1000000);
-    vt[i].y = (int)ceil (vertices[i].y * 1000000);
-    vt[i].z = (int)ceil (vertices[i].z * 1000000);
+    vt[i].x = (float)ceil (vertices[i].x * 1000000.0);
+    vt[i].y = (float)ceil (vertices[i].y * 1000000.0);
+    vt[i].z = (float)ceil (vertices[i].z * 1000000.0);
   }
 
   // First sort so that all (nearly) equal vertices are together.
