diff -urN orig/plugins/video/canvas/wxgl/GLWXDriver2D.cpp fix/plugins/video/canvas/wxgl/GLWXDriver2D.cpp
--- orig/plugins/video/canvas/wxgl/GLWXDriver2D.cpp	2009-08-08 22:01:32.000000000 -0500
+++ fix/plugins/video/canvas/wxgl/GLWXDriver2D.cpp	2009-08-10 22:16:44.000000000 -0500
@@ -65,6 +65,10 @@
 #include <GL/glx.h>
 #endif
 
+#ifdef CS_PLATFORM_MACOSX
+#import <mach-o/dyld.h>
+#endif
+
 #include "GLWXDriver2D.h"
 
 CS_IMPLEMENT_PLUGIN
@@ -133,6 +137,19 @@
   return (void*)wglGetProcAddress (funcname);
 #elif defined(USE_GLX)
     return (void*)glXGetProcAddressARB ((const GLubyte*) funcname);
+#elif defined(CS_PLATFORM_MACOSX)
+    // Get the address of a procedure (for OGL use.)
+    NSSymbol symbol;
+    csString symbolName;
+    // Prepend a '_' for the Unix C symbol mangling convention
+    symbolName << '_' << funcname;
+    if (NSIsSymbolNameDefined (symbolName))
+    {
+      symbol = NSLookupAndBindSymbol (symbolName);
+      return NSAddressOfSymbol (symbol);
+    }
+    else
+      return 0;
 #else
   return 0;
 #endif
@@ -244,6 +261,29 @@
       WX_GL_MIN_ACCUM_ALPHA, currentFormat[glpfvAccumAlphaBits],
       0
   };
+#elif defined (CS_PLATFORM_MACOSX)
+    currentFormat[glpfvColorBits] = GetPixelFormat()->PixelBytes * 8;
+    currentFormat[glpfvAlphaBits] = GetPixelFormat()->AlphaBits;
+    currentFormat[glpfvDepthBits] = (GetPixelFormat()->PixelBytes * 8) - GetPixelFormat()->AlphaBits;
+    currentFormat[glpfvStencilBits] = 8;
+    currentFormat[glpfvAccumColorBits] = GetPixelFormat()->RedShift + GetPixelFormat()->BlueShift + GetPixelFormat()->GreenShift;
+    currentFormat[glpfvAccumAlphaBits] = 0; //no equivalent according to native OSX canvas code
+    int desired_attributes[] =
+    {
+      WX_GL_RGBA,
+      WX_GL_DOUBLEBUFFER,
+      WX_GL_DEPTH_SIZE, currentFormat[glpfvDepthBits],
+      WX_GL_MIN_RED, GetPixelFormat()->RedBits,
+      WX_GL_MIN_BLUE, GetPixelFormat()->BlueBits,
+      WX_GL_MIN_GREEN, GetPixelFormat()->GreenBits,
+      WX_GL_MIN_ALPHA, currentFormat[glpfvAlphaBits],
+      WX_GL_STENCIL_SIZE, currentFormat[glpfvStencilBits],
+      WX_GL_MIN_ACCUM_RED, (GetPixelFormat()->RedMask / pow(2, GetPixelFormat()->RedShift) + 1) / 32,
+      WX_GL_MIN_ACCUM_BLUE, (GetPixelFormat()->BlueMask / pow(2, GetPixelFormat()->BlueShift) + 1) / 32,
+      WX_GL_MIN_ACCUM_GREEN, (GetPixelFormat()->GreenMask / pow(2, GetPixelFormat()->GreenShift) + 1) / 32,
+      WX_GL_MIN_ACCUM_ALPHA, currentFormat[glpfvAccumAlphaBits],
+      0
+    };
 #else
     int desired_attributes[] =
     {
